engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractors.java

```diff
@@ -50,23 +50,33 @@ public ValueExtractors(Iterable<ValueExtractor<?>> externalExtractors) {
 		tmpValueExtractors.add( new ValueExtractorDescriptor( LegacyIterableValueExtractor.INSTANCE ) );
 		tmpValueExtractors.add( new ValueExtractorDescriptor( IterableValueExtractor.INSTANCE ) );
 
+		tmpValueExtractors.add( new ValueExtractorDescriptor( OptionalValueExtractor.INSTANCE ) );
 		tmpValueExtractors.add( new ValueExtractorDescriptor( ObjectValueExtractor.INSTANCE ) );
 
 		valueExtractors = Collections.unmodifiableList( tmpValueExtractors );
 	}
 
 	public ValueExtractor<?> getCascadedValueExtractor(Class<?> valueType, TypeVariable<?> cascadingTypeParameter) {
-		Map<Class<?>, Map<TypeVariable<?>, TypeVariable<?>>> allBindings = TypeVariableBindings.getTypeVariableBindings( valueType );
+		Map<Class<?>, Map<TypeVariable<?>, TypeVariable<?>>> allBindings = null;
+
+		if ( cascadingTypeParameter != AnnotatedObject.INSTANCE && cascadingTypeParameter != ArrayElement.INSTANCE ) {
+			allBindings = TypeVariableBindings.getTypeVariableBindings( (Class<?>) cascadingTypeParameter.getGenericDeclaration() );
+		}
```

Comment:
I was talking about the `allBindings` map. I don't feel comfortable having it set to null in some cases while we are using it below without a null check. For the time being, I don't think we can be in the case where it is null and we need it in the for loop but I'd better be safe and not have this pattern.